File: solutions/python/two-bucket/1/two_bucket.py
from collections import deque
from math import gcd


def measure(bucket_one, bucket_two, goal, start_bucket):
    """Measure exact amount using two buckets with BFS.
    
    Args:
        bucket_one: Size of bucket one
        bucket_two: Size of bucket two
        goal: Desired amount to measure
        start_bucket: Which bucket to fill first ("one" or "two")
    
    Returns:
        Tuple of (num_actions, bucket_with_goal, other_bucket_amount)
    
    Raises:
        ValueError: If goal is impossible to reach
    """
    # Goal must be achievable - check using GCD property
    if goal > max(bucket_one, bucket_two):
        raise ValueError("Goal is larger than both buckets")
    
    if gcd(bucket_one, bucket_two) != 1 and goal % gcd(bucket_one, bucket_two) != 0:
        raise ValueError("Goal is not achievable with these bucket sizes")
    
    # Initial state: fill the starting bucket
    if start_bucket == "one":
        initial_state = (bucket_one, 0)
        start_id = "one"
    else:
        initial_state = (0, bucket_two)
        start_id = "two"
    
    # BFS to find shortest path
    queue = deque([(initial_state, 1)])  # (state, num_actions)
    visited = {initial_state}
    
    while queue:
        (b1, b2), actions = queue.popleft()
        
        # Check if goal reached
        if b1 == goal:
            return (actions, "one", b2)
        if b2 == goal:
            return (actions, "two", b1)
        
        # Generate all possible next states
        next_states = []
        
        # Action 1: Empty bucket one
        next_states.append((0, b2))
        
        # Action 2: Empty bucket two
        next_states.append((b1, 0))
        
        # Action 3: Fill bucket one
        next_states.append((bucket_one, b2))
        
        # Action 4: Fill bucket two
        next_states.append((b1, bucket_two))
        
        # Action 5: Pour bucket one into bucket two
        pour_amount = min(b1, bucket_two - b2)
        next_states.append((b1 - pour_amount, b2 + pour_amount))
        
        # Action 6: Pour bucket two into bucket one
        pour_amount = min(b2, bucket_one - b1)
        next_states.append((b1 + pour_amount, b2 - pour_amount))
        
        for state in next_states:
            if state not in visited:
                # Check the constraint: can't end with starting bucket empty and other full
                if start_bucket == "one" and state != (0, bucket_two):
                    visited.add(state)
                    queue.append((state, actions + 1))
                elif start_bucket == "two" and state != (bucket_one, 0):
                    visited.add(state)
                    queue.append((state, actions + 1))
    
    raise ValueError("Goal is not achievable")

File: solutions/ruby/two-bucket/1/two_bucket.rb
require 'set'

class TwoBucket
  attr_reader :moves, :goal_bucket, :other_bucket

  def initialize(size1, size2, goal, start)
    @size1 = size1
    @size2 = size2
    @goal = goal
    @start = start
    solve
  end

  private

  def solve
    start_state = @start == "one" ? [@size1, 0] : [0, @size2]
    visited = Set.new([start_state])
    queue = [[start_state, 1]]
    best = nil

    until queue.empty?
      (b1, b2), steps = queue.shift
      next if best && steps > best[:moves]

      result = check_goal(b1, b2, steps)
      if result
        if best.nil? || steps < best[:moves]
          best = result
        elsif steps == best[:moves]
          if @size1 >= @goal || @size2 >= @goal
            if result[:goal_bucket] == @start
              best = result
            end
          end
        end
        next
      end

      possible_moves(b1, b2).each do |state|
        next if visited.include?(state)
        visited.add(state)
        queue << [state, steps + 1]
      end
    end

    @moves = best[:moves]
    @goal_bucket = best[:goal_bucket]
    @other_bucket = best[:other_bucket]
  end

  def check_goal(b1, b2, steps)
    found = []
    found << { moves: steps, goal_bucket: "one", other_bucket: b2 } if b1 == @goal
    found << { moves: steps, goal_bucket: "two", other_bucket: b1 } if b2 == @goal
    return nil if found.empty?

    if @start == "one" && @size1 >= @goal
      found.find { |r| r[:goal_bucket] == "one" }
    elsif @start == "two" && @size2 >= @goal
      found.find { |r| r[:goal_bucket] == "two" }
    else
      found.first
    end
  end

  def possible_moves(b1, b2)
    moves = []
    moves << [@size1, b2] unless b1 == @size1
    moves << [b1, @size2] unless b2 == @size2
    moves << [0, b2] unless b1 == 0
    moves << [b1, 0] unless b2 == 0
    unless b1 == 0 || b2 == @size2
      pour = [b1, @size2 - b2].min
      moves << [b1 - pour, b2 + pour]
    end
    unless b2 == 0 || b1 == @size1
      pour = [b2, @size1 - b1].min
      moves << [b1 + pour, b2 - pour]
    end
    moves.uniq.reject { |state| invalid?(state) }
  end

  def invalid?(state)
    b1, b2 = state
    if @start == "one"
      b1 == 0 && b2 == @size2
    else
      b2 == 0 && b1 == @size1
    end
  end
end
